text/0000-cargo-libstd-awareness.md

```diff
@@ -0,0 +1,236 @@
+As followup work, it expected that packages (probably just the workspace root, definitely not non-packages like cargo config) will be able to specify the "default" source, i.e. the one used when none is specified (today this is always crates.io).
+Similarly, one could specify a "stdlib" source, to be used for `stdlib = true` deps instead of the compiler source binary registry or sysroot binary mock registry.
+This would simplify rustbuild as it could use that once instead of `[replace]` for each package.
+This doesn't require any planning from this RFC.
+
+More importantly, it would be nice to move stdlib crates that don't use unstable features to crates.io.
+`collections` and `test` almost don't use any unstable and are thus good candidates for this.
+With something like what is described in the first paragraph, it could be possible for individual packages to instruct Cargo to first check crates.io, and then the compiler source registry, for stdlib crates.
+But this shifts the burden to individual packages, and means we'd still need to vendor source of any crate moved to crates.io in the compiler source registry for packages that didn't make the switch.
+
+More interesting would be to change Cargo's *default* behavior to check both the compiler-specific sources (compiler source registry and sysroot binary mock registry) and crates.io.
+This would allow standard crates to seamlessly migrate to crates.io without extra work per package.
+This could be either be done where crates.io overrides the compiler-specific sources, or the compiler specific sources override crates.io.
+We don't want to commit to either variant in this RFC, however, so we instead want to keep all 3 options open (no fallback, crates.io over compiler-specific, compiler-specific over crates.io).
+To achieve this, we want to keep the sysroot/compiler source registry and crates.io disjoint: no package should be contained in both sources.
+That way unioning them together with either priority (the fallback scheme effectively crates a union source) has the same effect.
+
+The easiest way to achieve this is to make sure that standard library crates use names reserved on crates.io.
+We don't want to bake crates.io policy into Cargo however, so instead of absolutely prohibiting stdlib deps with non-reserved names, crates.io will just lint packages being uploaded.
+Also, care will be taken so that any stdlib crate that is stabalized must use a reserved name or already be published on crates.io.
+That still doesn't protect unpublished packages using unstable stdlib crates without reserved names from breakage, but due to their use of unstable interfaces we have no obligation to keep them working.
+Also, once we have an option to explicitly provide the source for stdlib deps, they can force the behavior they want.
+This seems good enough.
+
+
+# Drawbacks
+
+ - The mock registry for sysroot binaries is a disgusting hack.
```

Comment:
This characterization of the design as a disgusting hack is not so constructive. If it has drawbacks please state them plainly.